core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/general/PID1Rule.java

```diff
@@ -0,0 +1,96 @@
+import org.openjdk.jmc.flightrecorder.rules.IResultValueProvider;
+import org.openjdk.jmc.flightrecorder.rules.IRule;
+import org.openjdk.jmc.flightrecorder.rules.ResultBuilder;
+import org.openjdk.jmc.flightrecorder.rules.Severity;
+import org.openjdk.jmc.flightrecorder.rules.TypedResult;
+import org.openjdk.jmc.flightrecorder.rules.jdk.messages.internal.Messages;
+import org.openjdk.jmc.flightrecorder.rules.util.JfrRuleTopics;
+import org.openjdk.jmc.flightrecorder.rules.util.RulesToolkit;
+
+public class PID1Rule implements IRule {
+	private static final String PID1Rule_RESULT_ID = "PID1Rule";
+
+	private static final Map<String, RulesToolkit.EventAvailability> REQUIRED_EVENTS = RulesToolkit.RequiredEventsBuilder
+			.create().addEventType(JdkTypeIDs.VM_INFO, RulesToolkit.EventAvailability.AVAILABLE).build();
+
+	@Override
+	public String getId() {
+		return PID1Rule_RESULT_ID;
+	}
+
+	@Override
+	public String getTopic() {
+		return JfrRuleTopics.JVM_INFORMATION;
+	}
+
+	@Override
+	public String getName() {
+		return Messages.getString(Messages.PID1Rule_RULE_NAME);
+	}
+
+	@Override
+	public Map<String, RulesToolkit.EventAvailability> getRequiredEvents() {
+		return REQUIRED_EVENTS;
+	}
+
+	@Override
+	public RunnableFuture<IResult> createEvaluation(
+		IItemCollection items, IPreferenceValueProvider valueProvider, IResultValueProvider resultProvider) {
+		return new FutureTask<>(() -> getResult(items, valueProvider));
+	}
+
+	private IResult getResult(IItemCollection items, IPreferenceValueProvider valueProvider) {
+		IItemCollection vmInfoEvents = items.apply(ItemFilters.type(JdkTypeIDs.VM_INFO));
+		IAggregator<Set<String>, ?> distinct = Aggregators.distinct(JdkAttributes.PID);
+		Set<String> aggregate = vmInfoEvents.getAggregate(distinct);
+		if (aggregate == null || aggregate.isEmpty()) {
+			return ResultBuilder.createFor(this, valueProvider).setSeverity(Severity.IGNORE)
+					.setSummary(Messages.getString(Messages.PID1Rule_NO_PID)).build();
+		}
+
+		if (aggregate.size() > 1) {
+			String warningMessage = "There is more than one PID information available. This should not be possible."; //$NON-NLS-1$
+			Logger.getLogger(this.getClass().getName()).warning(warningMessage);
+			return RulesToolkit.getNotApplicableResult(this, valueProvider, warningMessage);
+		}
+
+		String pid = aggregate.iterator().next();
+		if ("1".equals(pid)) {
+			return ResultBuilder.createFor(this, valueProvider).setSeverity(Severity.WARNING)
+					.setSummary(Messages.getString(Messages.PID1Rule_TEXT_INFO))
+					.setExplanation(Messages.getString(Messages.PID1Rule_TEXT_INFO_LONG)).build();
+		}
+
+		return ResultBuilder.createFor(this, valueProvider).setSeverity(Severity.OK)
+				.setSummary(Messages.PID1Rule_TEXT_OK).build();
+	}
+
+	@Override
+	public Collection<TypedPreference<?>> getConfigurationAttributes() {
+		return Collections.emptyList();
+	}
+
+	@Override
+	public Collection<TypedResult<?>> getResults() {
+		return Collections.emptyList();
```

Comment:
configuration attributes and results should be empty in this case, or?

Reply:
Hi Hendrik!

Great progress on the rule! Configuration attributes should be empty, since the rule is not configurable. For the result, you should at least return the score attribute (we'll make the API nicer for JMC 10), and even though we know the PID is one when the rule is triggering, perhaps we should add the PID attribute to the results as well. So, following the pattern of other rules, perhaps add:

	private static final Collection<TypedResult<?>> RESULT_ATTRIBUTES = Arrays
			.<TypedResult<?>> asList(TypedResult.SCORE, PID);

Reply:
I think this is fine actually, if we don't expect the output to ever be dynamic then this I don't see the need to add a result attribute.

Reply:
Can we think of situations where we would ever want to use the pid in the result? For example highlighting the process in the processes list with the result (good or bad) etc?